hook-janitor/src/webhooks.rs

```diff
@@ -1,55 +1,454 @@
+                   END AS last_error,
+                   count(*) as failures
+            FROM {0}
+            WHERE status = 'failed'
+              AND queue = $1
+            GROUP BY hour, team_id, plugin_config_id, last_error
+            ORDER BY hour, team_id, plugin_config_id, last_error;
+            "#,
+            self.table_name
+        );
+
+        let rows = sqlx::query_as::<_, FailedRow>(&base_query)
+            .bind(&self.queue_name)
+            .fetch_all(&mut **tx)
+            .await
+            .map_err(|e| WebhookCleanerError::GetFailedRowsError { error: e })?;
+
+        Ok(rows)
+    }
+
+    fn serialize_failed_rows(&self, failed_rows: Vec<FailedRow>) -> Result<Vec<String>> {
+        let mut payloads = Vec::new();
+
+        for row in failed_rows {
+            let app_metric = AppMetric {
+                timestamp: row.hour,
+                team_id: row.team_id,
+                plugin_config_id: row.plugin_config_id,
+                job_id: None,
+                category: AppMetricCategory::Webhook,
+                successes: 0,
+                successes_on_retry: 0,
+                failures: row.failures,
+                error_uuid: Some(Uuid::now_v7()),
+                error_type: Some(row.last_error.r#type),
+                error_details: Some(row.last_error.details),
+            };
+
+            let payload = serde_json::to_string(&app_metric)
+                .map_err(|e| WebhookCleanerError::SerializeRowsError { error: e })?;
+
+            payloads.push(payload)
+        }
+
+        Ok(payloads)
+    }
+
+    async fn send_messages_to_kafka(&self, payloads: Vec<String>) -> Result<()> {
+        let mut delivery_futures = Vec::new();
+
+        for payload in payloads {
+            match self.kafka_producer.send_result(FutureRecord {
+                topic: self.app_metrics_topic.as_str(),
+                payload: Some(&payload),
+                partition: None,
+                key: None::<&str>,
+                timestamp: None,
+                headers: None,
+            }) {
+                Ok(future) => delivery_futures.push(future),
+                Err((error, _)) => return Err(WebhookCleanerError::KafkaProduceError { error }),
+            }
+        }
+
+        for result in join_all(delivery_futures).await {
+            match result {
+                Ok(Ok(_)) => {}
+                Ok(Err((error, _))) => {
+                    return Err(WebhookCleanerError::KafkaProduceError { error })
+                }
+                Err(_) => {
+                    // Cancelled due to timeout while retrying
+                    return Err(WebhookCleanerError::KafkaProduceCanceled);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn delete_observed_rows(&self, tx: &mut Transaction<'_, Postgres>) -> Result<u64> {
+        // This DELETE is only safe because we are in serializable isolation mode, see the note
+        // in `start_serializable_txn`.
```

Comment:
Wish there was a way to enforce this: Seems like we take a tx, but that can be any transaction, not necessarily one that we started with serializable isolation mode.

Reply:
This is a bit nitpicky, especially for a first implementation of cleanup when the risk of mixing up transactions is very low. Happy to leave this for later, although I am a big fan of checking stuff at compile time if possible.

Reply:
No, I think this is a good idea, and should be easy enough by just wrapping it in a simple struct for some protection.